### PR TITLE
SINGA-505 Computational graph with memory optimization

### PR DESCRIPTION
**Pull request by XJDKC**
# Overview
This PR adds the computational graph with memory optimization. It is based on the code developed by @chrishkchris and @XJDKC and some discussions with @nudles.

# Features
There are four main features in this PR, namely the construction of the computational graph, lazy allocation, automatic recycling, and synchronization pipeline. Details as follows:
* `Computational graph construction`: Construct a computational graph based on the user-defined neural network or expressions and then run the graph to accomplish the training task. The computational graph also includes operations like synch and fused synch in the communicator.
* `Lazy allocation`: When blocks need to be allocated, devices won't allocate memory for them immediately. Only when an operation uses this block for the first time, memory allocation will be performed.
* `Automatic recycling`: Automatically deallocate the intermediate tensors which won't be used again in the following operations when we are running the graph in an iteration.
* `Synchronization pipeline`: In previous synchronization operations, buffers were used to synchronize multiple tensors at once. But the communicator needs to collect all the tensors before copying them into the buffer. Synchronization pipeline can copy tensors to the buffer separately, which reduces the time for synchronous operations.

# Design
1. Computational graph construction
    * Use the technique of delayed execution to falsely perform operations in the forward propagation and backward propagation once. Buffer all the operations and the tensors read or written by each operation. 
    * Calculate dependencies between all the operations to decide the order of execution. (Support directed cyclic graph)
    * Execute all the operations in the order we just calculated to update all the parameters.
    * The system will only analyze the same graph once. If new operations are added to the graph, the calculation graph will be re-analyzed.
    * Provided a module class for users to use this feature more conveniently.
2. Lazy allocation
    * When a device needs to create a new block, just pass the device to that block instead of allocating a piece of memory from the mempool and passing the pointer to that block.
    * When the block is accessed for the first time, let the device corresponding to the block allocate memory and then access it.
3. Automatic recycling
    * When calculating dependencies between the operations during the graph construction, the reference count of tensors can also be calculated.
    * When an operation is completed, we can decrease the reference count of tensors the operation used.
    * If a tensor's reference count reaches zero, it means the tensor won't be accessed by latter operations and we can recycle its memory.
    * The program will track the usage of the block. If a block is used on the python side, it will not be recycled, which is convenient for debugging on the python side.
4. Synchronization pipeline
    * If a tensor needs fusion synchronization, it will be copied to the buffer immediately and don't need to gather all the tensors. Because the copy operation is advanced, it takes less time to do real synchronization. This optimizes the use of the GPU.

# Changes
* `Tensor`&`Operation`&`Communicator`
    * Change the capture type of tensors in lambda expressions to achieve delayed execution.
    * Change the type of input and output parameters to ensure that the input and output of the operation are tensors.
    * Submit synchronous operations to the device for execution.
* `Device`: Add code for 
    * Buffering operations
    * RunGraph
* `Scheduler`
    * Add operations in the graph to construct the graph.
    * Analyse the graph, calculating dependencies and deciding the reallocation of the blocks.
    * Execute the operations in the graph. 
* `Block`: Add a member variable of type device to help to do the lazy allocation. Add a function to help to do automatic recycling.
* `Communicator` and `opt.py`: Support the synchronization pipeline.
* `Swig`: add some interfaces
* `Module`: Provide a Module class on the python side for users to use the graph more conveniently.
*  `Examples`: Add some examples with operations buffering by using Module class.

# Evaluation
## Single node
* Experiment settings
    * Modle
      * dev branch: ResNet50 in [resnet.py](../tree/dev/examples/autograd/resnet.py)
      * this pr: ResNet50 in [resnet_module.py](https://github.com/XJDKC/singa/tree/Experiment_buff_ops/examples/autograd/resnet_module.py)
    * GPU: Nvidia RTX 2080Ti
* Explanation
    * `s` ：second
    * `it` ： iteration
    * `Memory`：peak memory usage of single GPU
    * `Throughout`：number of pictures processed per second
    * `Time`：total time
    * `Speed`：iterations per second
    * `Reduction`：the memory usage reduction rate compared with dev branch
    * `Seepdup`: speedup ratio compared with dev branch
* Result: 
<table style="text-align: center">
    <tr>
        <th style="text-align: center">Batchsize</th>
        <th style="text-align: center">Cases</th>
        <th style="text-align: center">Memory(MB)</th>
        <th style="text-align: center">Time(s)</th>
        <th style="text-align: center">Speed(it/s)</th>
        <th style="text-align: center">Throughput</th>
        <th style="text-align: center">Reduction</th>
        <th style="text-align: center">Speedup</th>
    </tr>
    <tr>
        <td rowspan="4">16</td>
        <td nowrap>dev</td>
        <td>4975</td>
        <td>14.1952</td>
        <td>14.0893</td>
        <td>225.4285</td>
        <td>0.00%</td>
        <td>1.0000</td>
    </tr>
    <tr>
        <td nowrap>PR:no graph</td>
        <td>4995</td>
        <td>14.1264</td>
        <td>14.1579</td>
        <td>226.5261</td>
        <td>-0.40%</td>
        <td>1.0049</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, bfs</td>
        <td>3283</td>
        <td>13.7438</td>
        <td>14.5520</td>
        <td>232.8318</td>
        <td>34.01%</td>
        <td>1.0328</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, serial</td>
        <td>3265</td>
        <td>13.7420</td>
        <td>14.5540</td>
        <td>232.8635</td>
        <td>34.37%</td>
        <td>1.0330</td>
    </tr>
    <tr>
        <td rowspan="4">32</td>
        <td nowrap>dev</td>
        <td>10119</td>
        <td>13.4587</td>
        <td>7.4302</td>
        <td>237.7649</td>
        <td>0.00%</td>
        <td>1.0000</td>
    </tr>
    <tr>
        <td nowrap>PR:no graph</td>
        <td>10109</td>
        <td>13.2952</td>
        <td>7.5315</td>
        <td>240.6875</td>
        <td>0.10%</td>
        <td>1.0123</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, bfs</td>
        <td>6839</td>
        <td>13.1059</td>
        <td>7.6302</td>
        <td>244.1648</td>
        <td>32.41%</td>
        <td>1.0269</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, serial</td>
        <td>6845</td>
        <td>13.0489</td>
        <td>7.6635</td>
        <td>245.2312</td>
        <td>32.35%</td>
        <td>1.0314</td>
    </tr>
</table>

## Multi processes
* Experiment settings
    * Model
      * dev branch: ResNet50 in [resnet_dist.py](../tree/dev/examples/autograd/resnet_dist.py)
      * this pr: ResNet50 in [resnet_module.py](https://github.com/XJDKC/singa/tree/Experiment_buff_ops/examples/autograd/resnet_module.py)
    * GPU: Nvidia RTX 2080Ti \* 2
    * MPI: two MPI processes on one node
* Explanation: the same as above
* Result: 
<table style="text-align: center">
    <tr>
        <th style="text-align: center">Batchsize</th>
        <th style="text-align: center">Cases</th>
        <th style="text-align: center">Memory(MB)</th>
        <th style="text-align: center">Time(s)</th>
        <th style="text-align: center">Speed(it/s)</th>
        <th style="text-align: center">Throughput</th>
        <th style="text-align: center">Reduction</th>
        <th style="text-align: center">Speedup</th>
    </tr>
    <tr>
        <td rowspan="4">16</td>
        <td nowrap>dev</td>
        <td>5439</td>
        <td>17.3323</td>
        <td>11.5391</td>
        <td>369.2522</td>
        <td>0.00%</td>
        <td>1.0000</td>
    </tr>
    <tr>
        <td nowrap>PR:no graph</td>
        <td>5427</td>
        <td>17.8232</td>
        <td>11.2213</td>
        <td>359.0831</td>
        <td>0.22%</td>
        <td>0.9725</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, bfs</td>
        <td>3389</td>
        <td>18.2310</td>
        <td>10.9703</td>
        <td>351.0504</td>
        <td>37.69%</td>
        <td>0.9725</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, serial</td>
        <td>3437</td>
        <td>17.0389</td>
        <td>11.7378</td>
        <td>375.6103</td>
        <td>36.69%</td>
        <td>1.0172</td>
    </tr>
    <tr>
        <td rowspan="4">32</td>
        <td nowrap>dev</td>
        <td>10547</td>
        <td>14.8635</td>
        <td>6.7684</td>
        <td>237.7649</td>
        <td>0.00%</td>
        <td>1.0000</td>
    </tr>
    <tr>
        <td nowrap>PR:no graph</td>
        <td>10503</td>
        <td>14.7746</td>
        <td>6.7684</td>
        <td>433.1748</td>
        <td>0.42%</td>
        <td>1.0060</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, bfs</td>
        <td>6935</td>
        <td>14.8553</td>
        <td>6.7384</td>
        <td>433.1748</td>
        <td>34.25%</td>
        <td>1.0269</td>
    </tr>
    <tr>
        <td nowrap>PR:with graph, serial</td>
        <td>7027</td>
        <td>14.3271</td>
        <td>6.9798</td>
        <td>446.7074</td>
        <td>33.37%</td>
        <td>1.0374</td>
    </tr>
</table>

From the table above, we can know:
* This PR does not affect training time and memory usage if the graph is disabled (has backward compatibility).
* This PR can significantly reduce memory usage and training time by using the graph.

# How to use
* An example of CNN:
```Python

class CNN(module.Module):

    def __init__(self, optimizer):
        super(CNN, self).__init__()

        self.conv1 = autograd.Conv2d(1, 20, 5, padding=0)
        self.conv2 = autograd.Conv2d(20, 50, 5, padding=0)
        self.linear1 = autograd.Linear(4 * 4 * 50, 500)
        self.linear2 = autograd.Linear(500, 10)
        self.pooling1 = autograd.MaxPool2d(2, 2, padding=0)
        self.pooling2 = autograd.MaxPool2d(2, 2, padding=0)

        self.optimizer = optimizer

    def forward(self, x):
        y = self.conv1(x)
        y = autograd.relu(y)
        y = self.pooling1(y)
        y = self.conv2(y)
        y = autograd.relu(y)
        y = self.pooling2(y)
        y = autograd.flatten(y)
        y = self.linear1(y)
        y = autograd.relu(y)
        y = self.linear2(y)
        return y

    def loss(self, x, ty):
        return autograd.softmax_cross_entropy(x, ty)

    def optim(self, loss):
        self.optimizer.backward_and_update(loss)

# initialization other objects
# ......
model = CNN(sgd)

# Train
for b in range(num_train_batch):
    # Generate the patch data in this iteration
    # ......

    # Copy the patch data into input tensors
    tx.copy_from_numpy(x)
    ty.copy_from_numpy(y)

    # Train the model
    out = model(tx)
    loss = model.loss(out, ty)
    model.optim(loss)
```
* More examples:
    * [MLP](https://github.com/XJDKC/singa/tree/Experiment_buff_ops/examples/autograd/mlp_module.py)
    * [CNN](https://github.com/XJDKC/singa/tree/Experiment_buff_ops/examples/autograd/cnn_module.py)
    * [ResNet](https://github.com/XJDKC/singa/tree/Experiment_buff_ops/examples/autograd/resnet_module.py)

# Plan
- [ ] Computation graph optimization: replace a subgraph of the input computation graph with another subgraph which is functionally equivalent to the original one. 
- [ ] Support recalculation and swapping out variables from the GPU.
- [ ] Perform operations in the graph in the order of DFS.
- [ ] Performing operations in parallel.

### CODE REVIEW

**Comment by chrishkchris** [comment]
please correct the result of b32 PR (no graph), that should be typo

**Comment by XJDKC** [comment]
Noted, thanks. I will fix it.

File: examples/autograd/mlp_buffer.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    loss = autograd.softmax_cross_entropy(x, target)
+    print("start backward")
+    for p, gp in autograd.backward(loss):
+        sgd.apply(0, gp, p, "")
+    dev.EnableGraph(False)
+
+    # exec the buffered ops
+    print("start executing buffered functions")
+    for i in range(1001):
+        dev.RunGraph()
```
[the remark this answers is not in the document]
Comment (XJDKC) [reply examples/autograd/mlp_buffer.py]: In this example, we just randomly generate the initial tensor and will not change it in the following training.

File: examples/autograd/mlp_buffer.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    print("w1:\n", tensor.to_numpy(w1))
+    print("b1:\n", tensor.to_numpy(b1))
+
+    sgd = optimizer.SGD(0.05)
+
+    # training process
+    print("start training")
+
+    # Buffer the operations
+    dev.EnableGraph(True)
```
[the remark this answers is not in the document]
Comment (XJDKC) [reply examples/autograd/mlp_buffer.py]: Now, we buffer all the operations in the graph and there is no separate buffer(i.e. buffer = graph).

File: examples/autograd/mnist_cnn_buffer.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    dev.EnableGraph(True)
+    autograd.training = True
+    out = model.forward(tx)
+    loss = autograd.softmax_cross_entropy(out, ty)
+    print("buffer softmax_cross_entropy")
+    for p, g in autograd.backward(loss):
+        sgd.update(p, g)
+        # print("update sgd")
+    autograd.training = False
+    dev.EnableGraph(False)
```
[the remark this answers is not in the document]
Comment (XJDKC) [reply examples/autograd/mnist_cnn_buffer.py]: I discussed with Chris yesterday, to support distributed training I need to change some code in  [communicator.cc](../tree/dev/src/io/communicator.cc)

File: examples/autograd/mlp_buffer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (XJDKC) [reply examples/autograd/mlp_buffer.py]: For real training, just call the copy_from_numpy as usual.

[the remark this answers is not in the document]
Comment (nudles) [reply examples/autograd/mlp_buffer.py]: copy the data to a placeholder variable in the loop?

File: examples/autograd/mnist_cnn_buffer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrishkchris) [reply examples/autograd/mnist_cnn_buffer.py]: > I discussed with Chris yesterday, to support distributed training I need to change some code in  communicator.cc

Yes, i.e. to buffer the all-reduce operation in communicator.cc, then it will support multi-process or MPI.
This is because the training is exactly the same as in single-gpu and multi-process (MPI), the only difference is that multi-process one has all-reduce.

File: examples/autograd/mlp_buffer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (XJDKC) [reply examples/autograd/mlp_buffer.py]: Yes, just like this.
```python
# Copy the patch data into input tensors
tx.copy_from_numpy(x)
ty.copy_from_numpy(y)
```

**Comment by nudles** [comment]
have you rebase/merged with the latest dev?

**Comment by XJDKC** [comment]
> have you rebase/merged with the latest dev?
> ready for merge?

I have rebased it before but did not rebase on the latest dev branch. I'll rebase my pr again and then this pr is ready for merge.
